### PR TITLE
LZ4 compression for pcap files

### PR DESCRIPTION
**Pull request by mfil**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

Describe changes:
- Add option to compress pcap log files with lz4 before writing to disk

Update to https://github.com/OISF/suricata/pull/3188, https://github.com/OISF/suricata/pull/3200 and https://github.com/OISF/suricata/pull/3206
- Fixed a bug where pcap logs in multi mode were always uncompressed
- Added ifdefs to build without liblz4 installed
- Fix Appveyor and Travis CI failures
- Fix issues pointed out by victorjulien in https://github.com/OISF/suricata/pull/3206

### CODE REVIEW

File: src/log-pcap.c [hunk 1]
```diff
@@ -1077,6 +1292,100 @@ static OutputInitResult PcapLogInitCtx(ConfNode *conf)
...
+            if (comp->pcap_buf_wrapper == NULL) {
+                SCLogError(SC_ERR_FOPEN, "SCFmemopen failed: %s",
+                        strerror(errno));
+                exit(EXIT_FAILURE);
+            }
+
+            /* Set lz4 preferences. */
+
+            memset(&comp->lz4f_prefs, '\0', sizeof(comp->lz4f_prefs));
+            comp->lz4f_prefs.frameInfo.blockSizeID = max4MB;
```
Comment (jasonish) [inline src/log-pcap.c]: This doesn't appear to be defined on newer versions of the lz4 lib, like on Fedora. Looks like this should be LZ4F_max4MB.

**Comment by jasonish** [comment]
Would it make sense to automatically add .lz4 as a file extension when lz4 compression is enabled? Perhaps default to having it added, but an option to remove it that could be in the docs, but left out of the default suricata.yaml?

Or is there a reason to not add the extension?

**Comment by jasonish** [comment]
Can you also merge some of these fixup commits into the commit they fix up?

File: src/log-pcap.c [hunk 2]
```diff
@@ -213,15 +240,58 @@ static int PcapLogCloseFile(ThreadVars *t, PcapLogData *pl)
...
 
-        if (pl->pcap_dumper != NULL)
+        if (pl->pcap_dumper != NULL) {
             pcap_dump_close(pl->pcap_dumper);
+#ifdef HAVE_LIBLZ4
+            PcapLogCompressionData *comp = &pl->compression;
+            if (comp->format == PCAP_LOG_COMPRESSION_FORMAT_LZ4) {
+                /* pcap_dump_close() has closed its output ``file'',
+                 * so the buffer has been freed - allocate it again. */
+                comp->pcap_buf = SCMalloc(comp->pcap_buf_size);
```
Comment (jasonish) [inline src/log-pcap.c]: pcap_dump_close doesn't free this buffer, so this leads to a memory leak that is picked up by ASAN. Perhaps all that is needed is a memset?

File: src/log-pcap.c [hunk 3]
```diff
@@ -213,15 +240,58 @@ static int PcapLogCloseFile(ThreadVars *t, PcapLogData *pl)
...
         if (pl->pcap_dead_handle != NULL)
             pcap_close(pl->pcap_dead_handle);
         pl->pcap_dead_handle = NULL;
 
+#ifdef HAVE_LIBLZ4
+        PcapLogCompressionData *comp = &pl->compression;
+        if (comp->format == PCAP_LOG_COMPRESSION_FORMAT_LZ4) {
+            /* pcap_dump_close did not write any data because we call
+             * pcap_dump_flush() after every write when writing
+	         * compressed output. */
```
Comment (jasonish) [inline src/log-pcap.c]: formatting

File: src/log-pcap.c [hunk 4]
```diff
@@ -320,18 +390,50 @@ static int PcapLogOpenHandles(PcapLogData *pl, const Packet *p)
...
+        else if (pl->compression.format == PCAP_LOG_COMPRESSION_FORMAT_LZ4) {
+            PcapLogCompressionData *comp = &pl->compression;
+            if ((pl->pcap_dumper = pcap_dump_fopen(pl->pcap_dead_handle,
+                    comp->pcap_buf_wrapper)) == NULL) {
+                SCLogError(SC_ERR_OPENING_FILE, "Error opening dump file %s",
+                        pcap_geterr(pl->pcap_dead_handle));
+                return TM_ECODE_FAILED;
+            }
+            comp->file = fopen(pl->filename, "w");
+            if (comp->file == NULL){
```
Comment (jasonish) [inline src/log-pcap.c]: space before {

**Comment by jasonish** [comment]
This feature should probably also have a feature ticket on redmine.

File: src/log-pcap.c (hunk 2 above)
Comment (mfil) [reply src/log-pcap.c]: Ouch, thanks for pointing that out! I thought that the buffer will always be freed if you close a pointer from fmemopen. But it seems that's not the case. It's only freed if you made fmemopen allocate the buffer itself by passing it a NULL pointer.

File: suricata.yaml.in [hunk 5]
```diff
@@ -376,6 +376,14 @@ outputs:
       # If set to a value will enable ring buffer mode. Will keep Maximum of "max-files" of size "limit"
       max-files: 2000
 
+      # Compression algorithm for pcap files. Possible values: none, lz4.
+      # Note that on Windows, enabling compression will *increase* disk I/O.
+      compression: none
+
+      # Further options for lz4 compression. The compression level can be set to a value between 0 and 16, where higher values mean higher compression, with 0 the default.
+      #lz4-checksum: no
+      #lz4-compression-level: 0
```
Comment (jasonish) [inline suricata.yaml.in]: I'm wondering if this should be simply lz4-level?

Comment (mfil) [reply suricata.yaml.in]: Sorry, I opened the next pull request before reading this...

**Pull request closed without merging**